packages/crayons-core/src/components/icon/icon.tsx

```diff
@@ -1,51 +1,155 @@
-import { Component, Prop, State, Watch, h } from '@stencil/core';
-
-// Icons Object
-import icons from './icon-assets/icons.json';
+import {
+  Build,
+  Component,
+  Element,
+  Prop,
+  State,
+  Watch,
+  h,
+} from '@stencil/core';
+import { fetchIcon } from '@freshworks/crayons-icon/icon-utils/icon-fetch-memoize';
 
 @Component({
   tag: 'fw-icon',
   styleUrl: 'icon.scss',
   assetsDirs: ['icon-assets'],
   shadow: true,
 })
+/**
+ * Class of the icon. Suppoprts Intersection Observer, Icon Cache, JS Imports for Tree shaking inline SVGs and SVGs from external CDNs or custom path within project.
+ */
 export class Icon {
   /**
-   * Identifier of the icon. The attribute’s value must be a valid svg file in the repo of icons (assets/icons).
+   * Identifier of the icon. The attribute’s value must be a valid svg file name or a JS Export of the svg file in the repo of icons (icon-assets/icons) or a named export of svg.
    */
   @Prop() name: string;
-
   /**
-   * Size of the icon, specified in number of  pixels.
-   * Default value is 12px defined using the --icon-size css variable.
+   * Size of the icon, specified in number of  pixels. This will be square coordinates of (w X h) = size X size
    */
-  @Prop() size: number;
-
+  @Prop() size = 16;
+  /**
+   * Root Margin in px or percentage for intersection observer
+   */
+  @Prop() xRootMargin = '50px';
+  /**
+   * Width of the icon, specified in number of  pixels.
+   */
+  @Prop() width = undefined;
+  /**
+   * Height of the icon, specified in number of  pixels.
+   */
+  @Prop() height = undefined;
   /**
    * Color in which the icon is displayed, specified as a standard CSS color or as a HEX code.
    */
   @Prop() color = '';
+  /**
+   * Enable Intersection Observer. Default is true.
+   */
+  @Prop() lazy = 'true';
+  /**
+   * Name of External Library to be used
+   */
+  @Prop() library = 'crayons';
+
+  @Element() el: HTMLElement;
+
+  @State() private visible = false;
+
+  private intersectionObserver: IntersectionObserver;
 
   @State() svgHTML = '';
 
-  @Watch('name')
-  private setSVGState(iconName: string) {
-    this.svgHTML = icons[iconName];
+  @State() ico_height = undefined;
+
+  @State() ico_width = undefined;
+
+  @State() eval_xOb = false;
+
+  @Watch('name') private async setSVGState(): Promise<void> {
+    const { name, visible, lazy, library } = this;
+    if (lazy === 'false') this.eval_xOb = true;
+    else if (lazy === 'true' || lazy === undefined) {
+      this.eval_xOb = visible;
+    }
+    if (!Build.isBrowser || !name || !this.eval_xOb) {
+      return;
+    }
+
+    if (
+      name.indexOf('svg') > -1 &&
+      name.indexOf('viewBox') > -1 &&
+      name.indexOf('fill') > -1
+    )
+      this.svgHTML = this.name;
+    else {
+      this.svgHTML = await fetchIcon(name, library);
+    }
+    this.ico_height = this.height !== undefined ? this.height : this.size;
```

Comment:
This will be a problem if we have the --icon-size CSS variable

Reply:
Fixed : --icon-size defaults | added props : crayon-icon | corrected typos | nolazy prop added.

Also Intersection Observer is showing great results and I have tested it number of times with various settings. There are examples in live in vuepress docs and results can be noted on browser. Many react icon libraries are providing similar implementation via useScreen hook which is using xOb only as implementation.